lib/fetch/util.js

```diff
@@ -200,8 +190,31 @@ function setRequestReferrerPolicyOnRedirect (request, actualResponse) {
 
   // 1. Let policy be the result of executing § 8.1 Parse a referrer policy
   // from a Referrer-Policy header on actualResponse.
-  // TODO:  https://w3c.github.io/webappsec-referrer-policy/#parse-referrer-policy-from-header
-  const policy = ''
+
+  // 8.1 Parse a referrer policy from a Referrer-Policy header
+  // 1. Let policy-tokens be the result of extracting header list values given `Referrer-Policy` and response’s header list.
+  const { headersList } = actualResponse
+  // 2. Let policy be the empty string.
+  // 3. For each token in policy-tokens, if token is a referrer policy and token is not the empty string, then set policy to token.
+  // 4. Return policy.
+  const policyHeader = (headersList.get('referrer-policy') ?? '').split(',')
+
+  // Note: As the referrer-policy can contain multiple policies
+  // separated by comma, we need to loop through all of them
+  // and pick the first valid one.
+  // Ref: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Referrer-Policy#specify_a_fallback_policy
+  let policy = ''
+  if (policyHeader.length > 0) {
+    // The right-most policy takes precedence.
+    // The left-most policy is the fallback.
+    for (let i = policyHeader.length; i !== 0; i--) {
```

Comment:
this seems to pick the last valid one?
ie. `same-origin, origin` will pick origin, when it should pick same-origin

Reply:
with the integrity option, the first valid one is always picked (assuming the algorithm "weight" is equal) so for example
`integrity: sha512-abc, sha512-def` will always choose sha512-abc assuming abc is a valid hash.

Reply:
>ie. same-origin, origin will pick origin, when it should pick same-origin

Yes the right-most

>with the integrity option, the first valid one is always picked (assuming the algorithm "weight" is equal) so for example
integrity: sha512-abc, sha512-def will always choose sha512-abc assuming abc is a valid hash.

I based the implementation on the MDN docs (see [ref](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Referrer-Policy#specify_a_fallback_policy)) which states:

>>`Referrer-Policy: no-referrer, strict-origin-when-cross-origin`
>>
>>In the above scenario, `no-referrer` is used only if the browser does not support the `strict-origin-when-cross-origin` policy.

Now accordingly to the spec Unknown Policy Values (ref: https://w3c.github.io/webappsec-referrer-policy/#unknown-policy-values), and following the _Example 18_:

>>Suppose older user agents don’t understand the ["unsafe-url"](https://w3c.github.io/webappsec-referrer-policy/#referrer-policy-unsafe-url) policy. A site can specify an ["origin"](https://w3c.github.io/webappsec-referrer-policy/#referrer-policy-origin) policy followed by an ["unsafe-url"](https://w3c.github.io/webappsec-referrer-policy/#referrer-policy-unsafe-url) policy: older user agents will ignore the unknown ["unsafe-url"](https://w3c.github.io/webappsec-referrer-policy/#referrer-policy-unsafe-url) value and use ["origin"](https://w3c.github.io/webappsec-referrer-policy/#referrer-policy-origin), while newer user agents will use ["unsafe-url"](https://w3c.github.io/webappsec-referrer-policy/#referrer-policy-unsafe-url) because it is the last to be processed.

The preferred policy should be appended at the very last of the list (right-most) while the fallbacks should be enlisted in order of priority (from right-to-left) being the left-most the final fallback to be chosen. This as the response on redirect can send the `Referrer Policy` header either independently or in a comma-separated list. Accordingly to _Example 19_:

>>To specify multiple policy values in the Referrer-Policy header, a site can send multiple Referrer-Policy headers:
`Referrer-Policy: no-referrer`
`Referrer-Policy: unsafe-url`
or, equivalently, multiple comma-separated header values:
>>
`Referrer-Policy: no-referrer,unsafe-url`

Where again, the right-most takes prescedence

Reply:
In example 18, `unsafe-url` isn't a supported policy so it is ignored. If it was known, it would be picked.

In 19 I don't see where the right most would be picked first, unless no-referrer wasn't a valid policy.

Reply:
The 19th was mostly as a reference and its theory was keeping the same logic of `unsafe-url` not being supported, though reading twice I think it doesn't 🤔 

For the 18th, after a third read, it is true that doesn't seem really clear whereas the order within the list, as this sentence makes it harder:

>older user agents will ignore the unknown ["unsafe-url"](https://w3c.github.io/webappsec-referrer-policy/#referrer-policy-unsafe-url) value and use ["origin"](https://w3c.github.io/webappsec-referrer-policy/#referrer-policy-origin), while newer user agents will use ["unsafe-url"](https://w3c.github.io/webappsec-referrer-policy/#referrer-policy-unsafe-url) because it is the last to be processed.

Which I understand as: `origin, unsafe-url`, so still quite ambiguous.
Let me move forward with testing how Chrome does it and change it accordingly 👍

Reply:
After doing the test on Chrome, it seems the first assumption was correct, it will take precedence over the right-most token on the `Referrer-Policy` header, and fallback to the `left-most` progressively.

I tested two scenarios with the same token policies, `no-referrer` and `origin` coming as responses from the Response on Redirect.

### Redirect response with `Referrer-Policy: origin,no-referrer`:

**Redirect Request/Response**
<img width="581" alt="image" src="https://user-images.githubusercontent.com/10200776/198876233-b721afbf-7d89-411e-9249-e9d939cc681c.png">

**Final Request/Response**
<img width="581" alt="image" src="https://user-images.githubusercontent.com/10200776/198876245-f48ae27c-f99e-47e6-be25-2165e1310b55.png">

**Result**: `no-referrer` took precede as being the left-most option (first to be processed)


### Redirect response with `Referrer-Policy: no-referrer,origin`:

**Redirect Request/Response**
<img width="581" alt="image" src="https://user-images.githubusercontent.com/10200776/198876327-8f598ef0-078d-4e84-9b7a-06bd4ce3f55e.png">


**Final Request/Response**
<img width="583" alt="image" src="https://user-images.githubusercontent.com/10200776/198876341-297a24d4-19cb-4f4d-82e8-a44002bfc42e.png">


**Result**: `origin` took precede as being the left-most option (first to be processed)

Based on Chrome, I think is possible to move forward with the current behavior, or wdyt? 🙂

Reply:
In firefox it is the left most, https://github.com/mozilla/gecko-dev/blob/76ccfc801e6b736c844cde3fddeab7a748fc8515/dom/security/ReferrerInfo.cpp#L165

Having it the right most goes against every other part of the spec and just generally doesn't make sense

Reply:
https://source.chromium.org/chromium/chromium/src/+/main:net/url_request/redirect_info.cc;l=50;bpv=0;bpt=1
https://github.com/mozilla/gecko-dev/blob/76ccfc801e6b736c844cde3fddeab7a748fc8515/dom/security/ReferrerInfo.cpp#L165

You seem to be correct; it chooses the last valid policy.

Specifically, the spec says `the value of the latest one will be used`.